### PR TITLE
Use `overrides` for a single `.eslintrc.js`.

### PR DESCRIPTION
**Pull request by rwjblue**
Update the `.eslintrc.js` used by apps and addons, to avoid needing separate "sprinkles" of eslint configuration throughout the projects codebase.

Compatible with `eslint@4.1.0`.

Additional reading:

* https://eslint.org/docs/user-guide/configuring#configuration-based-on-glob-patterns
* https://github.com/eslint/eslint/pull/8081

### CODE REVIEW

**Comment by rwjblue** [comment]
I will follow up and make sure that `ember-cli-eslint`'s blueprint is in sync once we settle on this...

File: blueprints/app/files/.eslintrc.js [hunk 1]
```diff
@@ -2,12 +2,39 @@ module.exports = {
   root: true,
   parserOptions: {
     ecmaVersion: 2017,
-    sourceType: 'module'
+    sourceType: 'module',
```
Comment (kellyselden) [inline blueprints/app/files/.eslintrc.js]: seems strange to only have trailing commas in this file. I vote for removal.

Comment (rwjblue) [reply blueprints/app/files/.eslintrc.js]: 😭 but ok 😄

Comment (Turbo87) [reply blueprints/app/files/.eslintrc.js]: I agree, we should add them to all files...

File: blueprints/app/files/.eslintrc.js [hunk 2]
```diff
@@ -2,12 +2,39 @@ module.exports = {
...
   rules: {
-  }
+  },
+  overrides: [
+    // node files
+    {
+      files: [
+        './ember-cli-build.js',
+        './testem.js',
+        './index.js',
```
Comment (Turbo87) [inline blueprints/app/files/.eslintrc.js]: should we use `./*.js` instead?

File: blueprints/app/files/.eslintrc.js [hunk 3]
```diff
@@ -2,12 +2,39 @@ module.exports = {
...
+        'config/**/*.js',
+      ],
+      parserOptions: {
+        sourceType: 'script',
+        ecmaVersion: 2015,
+      },
+      env: {
+        browser: false,
```
Comment (Turbo87) [inline blueprints/app/files/.eslintrc.js]: I think this has to be `browser: null` or something like that. iirc `true/false` only defines whether the global can be overwritten or something like that

File: blueprints/app/files/.eslintrc.js [hunk 4]
```diff
@@ -2,12 +2,39 @@ module.exports = {
...
+        node: true,
+      }
+    },
+
+    // test files
+    {
+      files: ['tests/**/*.js'],
+        env: {
```
Comment (Turbo87) [inline blueprints/app/files/.eslintrc.js]: indentation

File: blueprints/app/files/.eslintrc.js [hunk 5]
```diff
@@ -2,12 +2,39 @@ module.exports = {
...
-  }
+  },
+  overrides: [
+    // node files
+    {
+      files: [
+        './ember-cli-build.js',
+        './testem.js',
+        './index.js',
+        'config/**/*.js',
```
Comment (Turbo87) [inline blueprints/app/files/.eslintrc.js]: what about `tests/dummy/config/**/*.js`?

File: blueprints/app/files/.eslintrc.js [hunk 6]
```diff
@@ -2,12 +2,39 @@ module.exports = {
...
+      },
+      env: {
+        browser: false,
+        node: true,
+      }
+    },
+
+    // test files
+    {
+      files: ['tests/**/*.js'],
```
Comment (Turbo87) [inline blueprints/app/files/.eslintrc.js]: can we exclude `tests/dummy/`?

it's okay if we can't since we haven't done it before either, but would be nice though

File: blueprints/app/files/.eslintrc.js (hunk 1 above)
Comment (rwjblue) [reply blueprints/app/files/.eslintrc.js]: Ya, agreed. Though I’m happy to do this separately.

File: blueprints/app/files/.eslintrc.js (hunk 2 above)
Comment (rwjblue) [reply blueprints/app/files/.eslintrc.js]: Interesting. I hadn’t thought of that. Should work nicely.

File: blueprints/app/files/.eslintrc.js (hunk 3 above)
Comment (rwjblue) [reply blueprints/app/files/.eslintrc.js]: Really!?!?! Weird! I’ll look it up. I have used this on a few other projects, will test and update as needed.

File: blueprints/app/files/.eslintrc.js (hunk 5 above)
Comment (rwjblue) [reply blueprints/app/files/.eslintrc.js]: 👍 - good idea

File: blueprints/app/files/.eslintrc.js (hunk 3 above)
Comment (Turbo87) [reply blueprints/app/files/.eslintrc.js]: I just realized I might be confusing this with the stuff in the `globals` object, but we should verify just in case.

Comment (rwjblue) [reply blueprints/app/files/.eslintrc.js]: Ya, I was looking through the docs on configuring and I think you are thinking of the globals stuff.  For easy cross referencing:

* https://eslint.org/docs/user-guide/configuring#specifying-environments <- Docs on these `env` flags (no discussion of clobbering)
* https://eslint.org/docs/user-guide/configuring#specifying-globals <- Docs on `globals` (includes discussion on `true` / `false` flag to allow clobbering)

**Comment by kellyselden** [comment]
I would like to hold off merging until our meeting. I'm not 100% in favor of this. I like getting rid of the first line sprinkles of node files, but I'm not convinced that removing our tests/.eslintrc.js is a good idea. Everybody's documentation and blueprint injection touch that file (globals and test helpers), and it seems to me that it is in it's correct place.

**Comment by rwjblue** [comment]
> I would like to hold off merging until our meeting.

Seems good, happy to discuss...

> I like getting rid of the first line sprinkles of node files, but I'm not convinced that removing our tests/.eslintrc.js is a good idea. 

This is where we disagree. The current system of having many `.eslintrc` files all over makes it very difficult to understand what the current ruleset is at any given point (a number of times I've resorted to using `eslint --print-config some/file/path.js` just to figure out what the heck is running). Another issue is that (other than simple things like `env`) its not possible by adding a `.eslintrc` file to handle directories that contain _mixed_ rulesets .

> Everybody's documentation and blueprint injection touch that file (globals and test helpers), and it seems to me that it is in it's correct place.

I think this is actually a separate concern from the others. We should choose the right set of defaults, and recommend them while also providing time and assistance to addons that may need to make modifications. Modifying a single file (`.eslintrc.js`) in general, is easier than having to modify many potential locations...

**Comment by kellyselden** [comment]
We decided in the core meeting that this is the way forward. With module unification, test files will be strewn about a project, so wildcard paths inside a single eslint config is the future (TM).

File: blueprints/app/files/.eslintrc.js (hunk 6 above)
Comment (rwjblue) [reply blueprints/app/files/.eslintrc.js]: Yep, added and confirmed.  Works nicely!

**Comment by rwjblue** [comment]
Rebased and addressed all outstanding comments...

File: blueprints/app/files/.eslintrc.js (hunk 2 above)
Comment (rwjblue) [reply blueprints/app/files/.eslintrc.js]: Sadly, this doesn't work.  The main issue is that in the implementation (https://github.com/eslint/eslint/pull/8081) when the `files` and `excludedFiles` are passed in to `minimatch` ([here](https://github.com/eslint/eslint/blob/8e1a095f4cc3b371a0a424b66e13ec77734193a2/lib/config/config-ops.js#L368-L382)) they also pass `{ matchBase: true }` as options. 

From the `minimatch` README about `matchBase`:

> If set, then patterns without slashes will be matched against the basename of the path if it contains slashes. For example, a?b would match the path /xyz/123/acb, but not /xyz/acb/123.

**tl;dr** Due to the usage of `matchBase` it is not possible to use globs to match files in the project root. If we use `*.js` it matches _all_ files in the repo, if we use `./*.js` it matches no files, if we use `/*.js` it throws an error....

😭 😭 😭

**Comment by Turbo87** [comment]
@rwjblue this shouldn't block, but can you verify the behavior for in-repo-addons?

**Comment by rwjblue** [comment]
@Turbo87 - Good point, I will do some testing.  I _believe_ that we might need to add `lib/*/index.js` to the "node" config...
